python/ray/util/sgd/v2/backends/backend.py

```diff
@@ -102,9 +109,13 @@ def start(self, initialization_hook: Optional[Callable[[], None]] = None):
         self.worker_group = WorkerGroup(self._num_workers,
                                         self._num_cpus_per_worker,
                                         self._num_gpus_per_worker)
```

Comment:
I'm looking at the `Trainer._start_training` code and trying to understand if there are any potential issues due to the cyclic logic between `_start_training` and `_run_with_error_handling`. Besides a potentially hard to debug stack, one thing I'm wondering is if this will cause a deadlock when trying to start a second `WorkerGroup`.

Reply:
The cyclic logic should be handled when we eventually reach the maximum number of failures. There are tests checking for this: `test_max_failures` and `test_start_max_failure`.

Regarding the stack trace, that's a good point. I changed it so that if we hit the max failures, we raise the `RuntimeError` from None so users are not exposed to the `TraininWorkerError`. The actual error message from the training function itself is being logged here: https://github.com/ray-project/ray/pull/18090/files#diff-719dd761cccd53aaba210f04b4e509ed7ef14f63862c0e74cd12379ef1e7e2bfR35